### DISCUSSION TITLE
Volume.widget config

### DISCUSSION THREAD

**Discussion by guzz46** [LBCrion/sfwbar#192] [Q&A]
Does anyone know how to configure the volume.widget? I assume that's the widget that allows per app volume control? I've added it to my sfwbar.config and created a volume.widget in ~/.config/sfwbar/ but all I get is an icon, and when I click on the icon all I get is a popup saying sinks and clients, but both are empty, thanks.

**Comment by LBCrion** [comment]
Are you running the latest git version of sfwbar? Also, did you install sfwbar or compiled it locally and running it from the build directory?
If you've installed the git version, can you please try running `sfwbar -d -g "trigger.*volume-conf"` from a terminal?

**Reply by guzz46** [reply]
I installed it through the nixos package manager, and the version number is 1.0_beta13, and running sfwbar -d -g "trigger.*volume-conf" gives me (sfwbar:115018): Gtk-WARNING **: 19:03:05.166: Theme parsing error: <data>:1:39: '' is not a valid color name.
Do I need to be running the git version?
Edit:Looks like I need beta 14, so I guess I'm out of luck unless I figure out how to build it from source on nixos, the overlay I use for labwc doesn't seem to work on sfwbar, I get an error about not knowing how to unpack the source archive.

**Comment by LBCrion** [comment]
This would be because beta13 doesn't have the necessary module support. The latest volume.widget requires the current git version of sfwbar. The widget itself provides the layout and handling of the incoming information, but the information is sent by the module and that's only available in the git version.
I was hoping to release the new version last week, but discovered a nasty bug, which I just managed to fix. Hopefully I'll be able to get beta15 out later this week or next week.

**Reply by guzz46** [reply]
I've updated to beta_14 and the volume widget is working great thanks, I'm nit picking a bit but in the picture below is there a way to add an icon to the left of Rembrandt? or remove one of the semicolons? the one on the left mutes the client, and the one on the right opens separate channels, I don't really need the separate channels semicolon.

![240618_11-07-09](https://github.com/LBCrion/sfwbar/assets/98864413/c77348c0-d92a-44e2-b89a-3af7dd7bbeb9)

**Comment by LBCrion** [comment]
Are you using the default pulse.widget with beta14? The icon should be controlled by the CSS element image#pulse_volume_mute

**Reply by guzz46** [reply]
I'm using the volume.widget from here https://github.com/LBCrion/sfwbar/blob/main/config/volume.widget
I don't see a pulse.widget, I only see a wbar-pulse.widget

**Comment by LBCrion** [comment]
Are you sure this widget is actually used? I.e. is it included from your
main config? It shouldn't work since volume.widget uses the new unified
'volume' interface and interfaces have been introduced only after beta14!

**Reply by guzz46** [reply]
This is my config, I'm using the latest git version

```function("SfwbarInit") {
  SetLayer "top"
}

layout {

  button {
    action = "nwggrid -d /home/gareth/.local/share/applications/"                       
    value = "nix-snowflake-white"                          
    css = "* { min-height: 0.85cm; min-width: 0.85cm; border: 0px solid @theme_fg_color }"
  }

  button {
    action = "brave"                        
    value = "brave"                         
    css = "* { min-height: 0.85cm; min-width: 0.85cm; border: 0px solid @theme_fg_color }"
  }

  button {
    action = "thunar"                       
    value = "thunar"                         
    css = "* { min-height: 0.85cm; min-width: 0.85cm; border: 0px solid @theme_fg_color }"
  }

  button {
    action = "xfce4-terminal"                        
    value = "xfce4-terminal"                           
    css = "* { min-height: 0.85cm; min-width: 0.85cm; border: 0px solid @theme_fg_color }"
  }

  button {
    action = "mousepad"                        
    value = "mousepad"                          
    css = "* { min-height: 0.85cm; min-width: 0.85cm; border: 0px solid @theme_fg_color }"
  }

  button {
    action = "galculator"                      
    value = "galculator"                          
    css = "* { min-height: 0.85cm; min-width: 0.85cm; border: 0px solid @theme_fg_color }"
  }

  button {
    action = "/home/gareth/.config/scripts/steam"                       
    value = "/home/gareth/.icons/breeze-extra-gnome/apps/48/steam.png"                           
    css = "* { min-height: 0.85cm; min-width: 0.85cm; border: 0px solid @theme_fg_color }"
  }

  button {
    action = "kodi"                        
    value = "/home/gareth/.icons/breeze-extra-gnome/apps/48/kodi.png"                          
    css = "* { min-height: 0.85cm; min-width: 0.85cm; border: 0px solid @theme_fg_color }"
  }

  taskbar {
    css = "* { -GtkWidget-vexpand: false; border: }" 
    sort = false
    icons = true     
    labels = false   
    cols = 1
    action[2] = Close
    action[3] = Minimize
  }

  label {
    css = "* { -GtkWidget-vexpand: true; }"
  }

  include("cpu.widget")
  include("cpu-temp.widget")
  include("cpu-usage.widget")
  include("spacer.widget")
  include("gpu.widget")
  include("gpu-temp.widget")
  include("gpu-usage.widget")
  include("spacer.widget")
  include("ram.widget")
  include("ram-usage.widget")
  include("ram-percentage.widget")
  include("network-module.widget")
  include("volume.widget")

  include("clock.widget")
}

#CSS
window {
  -GtkWidget-direction: left;
}

button#taskbar_normal *, button#taskbar_active *, button#taskbar_normal:hover * {
  min-height: 1cm; 
  padding: 0.1mm;
  -GtkWidget-hexpand: true;
}


button#taskbar_normal , button#taskbar_active , button#taskbar_normal:hover {
 min-height: 0.75cm;
  padding: 0.5mm;
  background-image: none;
  border-radius: 0;
  border-image: none;
}

button#taskbar_active {
  background-color: #bbddff;
}

button#taskbar_normal:hover {
  background-color: #cccccc;
}

grid#layout {
  padding: 0.25mm;
  -GtkWidget-direction: bottom;
}

tooltip label {
  font: 0.4cm Sans;
  color: #ffffff;
}

**Comment by LBCrion** [comment]
Ah! Sorry, I thought you were on beta14. Indeed, git volume.config was
missing a couple of CSS properties. I updated it in the current git tree.

**Reply by guzz46** [reply]
My bad, I said I updated to beta_14, I should have mentioned I updated to the latest git, I didn't realize the git version was different from beta_14

**Reply by guzz46** [reply]
The new config works great thanks

![240619_09-17-22](https://github.com/LBCrion/sfwbar/assets/98864413/4d605382-6a49-44e7-af84-db4e126925a7)

**Reply by guzz46** [reply]
Also for anyone who doesn't need the channel expander you can comment out the following lines like so in the volume.widget

```
#    image '@DEVICE_ID@_expand' {
#      value = 'pan-end'
#      style = 'volume_control_expander'
#      tooltip = 'expand/collapse channels'
#      action = Function 'XVolumeToggleChannels'
#      loc(2,1,1,1)
#    }
```
![240619_10-40-31](https://github.com/LBCrion/sfwbar/assets/98864413/b26a707a-cb15-4181-99c6-da2196ba96fa)
